src/main.rs

```diff
@@ -76,63 +76,140 @@ impl Machine {
                 0x12 => {
-                    let dst = self.memory[self.pc + 1];
+                    let dst = self.memory[self.pc.wrapping_add(1)];
                     self.memory[dst] = self.acc;
                     2
                 }
 
+                // store immediate
+                0x13 => {
+                    let val = self.memory[self.pc.wrapping_add(1)];
+                    let dst = self.memory[self.pc.wrapping_add(2)];
+                    self.memory[dst] = val;
+                    3
+                }
+
+                // move
+                0x14 => {
+                    let src = self.memory[self.pc.wrapping_add(1)];
+                    let dst = self.memory[self.pc.wrapping_add(2)];
+                    self.memory[dst] = self.memory[src];
+                    3
+                }
+
+                // ==========================
+                // ========== 0x2_ ==========
+                // ==========================
+
                 // jump
                 0x20 => {
-                    self.pc = self.memory[self.pc + 1];
+                    self.pc = self.memory[self.pc.wrapping_add(1)];
                     0
                 }
 
                 // jump zero
                 0x21 => {
                     if self.acc == 0 {
-                        self.pc = self.memory[self.pc + 1];
+                        self.pc = self.memory[self.pc.wrapping_add(1)];
                         0
                     } else {
                         2
                     }
                 }
 
+                // ==========================
+                // ========== 0x3_ ==========
+                // ==========================
+
+                // add
+                0x30 => {
+                    let src = self.memory[self.pc.wrapping_add(1)];
+                    self.acc = self.acc.wrapping_add(self.memory[src]);
+                    2
+                }
+
                 // add immediate
                 0x31 => {
-                    self.acc = self.acc.wrapping_add(self.memory[self.pc + 1]);
+                    self.acc = self.acc.wrapping_add(self.memory[self.pc.wrapping_add(1)]);
+                    2
+                }
+
+                // subtract
+                0x32 => {
+                    let src = self.memory[self.pc.wrapping_add(1)];
+                    self.acc = self.acc.wrapping_sub(self.memory[src]);
                     2
                 }
 
                 // substract immediate
                 0x33 => {
-                    self.acc = self.acc.wrapping_sub(self.memory[self.pc + 1]);
+                    self.acc = self.acc.wrapping_sub(self.memory[self.pc.wrapping_add(1)]);
                     2
                 }
 
+                // shift right
+                0x34 => {
+                    self.acc = self.acc >> 1;
+                    1
+                }
+
+                // shift left
+                0x35 => {
+                    self.acc = self.acc << 1;
```

Comment:
I'm not sure if the Rust shift semantics match the one required by SHiT-CPU. I actually couldn't find information on this and thus [asked on StackOverflow](https://stackoverflow.com/q/51571066/2408867). But from simple tests it seems like zeroes are shifted in, like the SHiT-CPU requires. So I guess we can merge this :P

Reply:
This needs no fix (see [StackOverflow](https://stackoverflow.com/a/51571371/1752287)).